### PR TITLE
Add visibility change listener for performance telemetry

### PR DESCRIPTION
**Pull request by bmahall**
This PR-

Adds a boolean(visibilityChange) to performance data to indicate when Page Visibility changes during the AcquireTokenSilent, AcquireTokenByCode, SsoSilent calls. 
Tests included.

### CODE REVIEW

File: change/@azure-msal-browser-59aee50c-dfa5-47b3-aba3-935707c03d9f.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
+  "comment": "ATS- event vis change",
```
Comment (peterzenz) [inline change/@azure-msal-browser-59aee50c-dfa5-47b3-aba3-935707c03d9f.json]: Please remove the abbreviations here and throughout the code for readability.
`Adding page visibility change event for silent calls like AcquireTokenSilent`

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 2]
```diff
@@ -155,6 +157,7 @@ export abstract class ClientApplication {
 
         // Initialize the token cache
         this.tokenCache = new TokenCache(this.config, this.browserStorage, this.logger, this.browserCrypto);
+        this.trackPageVisibilityWithMeas = this.trackPageVisibilityWithMeas.bind(this);
```
Comment (peterzenz) [inline lib/msal-browser/src/app/ClientApplication.ts]: WithMeas => `WithMeasurement`  Consistency is good and you're mixing Meas as an abbreviation along with spelling it out completely like you do on lines 86, 87 just above.

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 3]
```diff
@@ -415,6 +418,31 @@ export abstract class ClientApplication {
         });
     }
 
+    private trackPageVisibilityWithMeas():void {
+        if(!this.ssoSilentMeasurement && !this.acquireTokenByCodeAsyncMeasurement) return;
+        this.logger.info("Perf: Event- visibility change detected" );
```
Comment (peterzenz) [inline lib/msal-browser/src/app/ClientApplication.ts]: Throughout this method you're mixing vis and visibility, please adjust to the long form of Visible or Visibility throughout your change.

File: lib/msal-browser/src/app/PublicClientApplication.ts [hunk 4]
```diff
@@ -25,6 +25,7 @@ export class PublicClientApplication extends ClientApplication implements IPubli
 
     // Active requests
     private activeSilentTokenRequests: Map<string, Promise<AuthenticationResult>>;
+    private astsAsyncMeasurement?: InProgressPerformanceEvent = undefined;
```
Comment (peterzenz) [inline lib/msal-browser/src/app/PublicClientApplication.ts]: What did the asts stand for?  @tnorling this seems to be from an older change.

File: lib/msal-browser/test/app/PublicClientApplication.spec.ts [hunk 5]
```diff
@@ -2656,6 +2812,49 @@ describe("PublicClientApplication.ts Class Unit Tests", () => {
...
+            sinon.stub(CryptoOps.prototype, "createNewGuid").returns(RANDOM_TEST_GUID);
+            const callbackId = pca.addPerformanceCallback((events => {
+                expect(events[0].correlationId).toBe(RANDOM_TEST_GUID)
+                expect(events[0].success).toBe(true);
+                expect(events[0].fromCache).toBe(false);
+                expect(events[0].accessTokenSize).toBe(16);
+                expect(events[0].idTokenSize).toBe(12);
+                expect(events[0].isNativeBroker).toBe(undefined);
+                expect(events[0].requestId).toBe(undefined);
+                expect(events[0].visChange).toBe(false);
```
Comment (peterzenz) [inline lib/msal-browser/test/app/PublicClientApplication.spec.ts]: Is it possible to get a test where this is true?

**Review by peterzenz — CHANGES_REQUESTED** [review]
I have a few comments on abbreviations that I'd like to see adjusted to be the longer for text for readability and consistency.  I'd also like to see if we can do a test where the page visibility changed datapoint is successfully set to true at the end of the test.

File: lib/msal-browser/src/app/PublicClientApplication.ts (hunk 4 above)
Comment (tnorling) [reply lib/msal-browser/src/app/PublicClientApplication.ts]: Not sure...probably a typo and was supposed to be "ATS" for acquireTokenSilent?

Comment (bmahall) [reply lib/msal-browser/src/app/PublicClientApplication.ts]: @peterzenz Thanks for pointing that out. Will create a separate PR for pre-existing typos.

File: lib/msal-browser/test/app/PublicClientApplication.spec.ts (hunk 5 above)
Comment (bmahall) [reply lib/msal-browser/test/app/PublicClientApplication.spec.ts]: updated.

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 6]
```diff
@@ -83,6 +83,8 @@ export abstract class ClientApplication {
     // Flag representing whether or not the initialize API has been called and completed
     protected initialized: boolean;
 
+    private ssoSilentMeasurement?: InProgressPerformanceEvent = undefined;
```
Comment (tnorling) [inline lib/msal-browser/src/app/ClientApplication.ts]: The '?' takes care of this for you
```suggestion
    private ssoSilentMeasurement?: InProgressPerformanceEvent;
```

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 7]
```diff
@@ -83,6 +83,8 @@ export abstract class ClientApplication {
...
+    private acquireTokenByCodeAsyncMeasurement?: InProgressPerformanceEvent = undefined;
```
Comment (tnorling) [inline lib/msal-browser/src/app/ClientApplication.ts]:
```suggestion
    private acquireTokenByCodeAsyncMeasurement?: InProgressPerformanceEvent;
```

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 8]
```diff
@@ -155,6 +157,7 @@ export abstract class ClientApplication {
 
         // Initialize the token cache
         this.tokenCache = new TokenCache(this.config, this.browserStorage, this.logger, this.browserCrypto);
+        this.trackPageVisibilityWithMeasurement = this.trackPageVisibilityWithMeasurement.bind(this);
```
Comment (tnorling) [inline lib/msal-browser/src/app/ClientApplication.ts:161]: What's the purpose of this?

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 9]
```diff
@@ -415,6 +418,31 @@ export abstract class ClientApplication {
         });
     }
 
+    private trackPageVisibilityWithMeasurement():void {
+        if(!this.ssoSilentMeasurement && !this.acquireTokenByCodeAsyncMeasurement) return;
```
Comment (tnorling) [inline lib/msal-browser/src/app/ClientApplication.ts]: More readable
```suggestion
        if(!this.ssoSilentMeasurement && !this.acquireTokenByCodeAsyncMeasurement) {
            return;
        }
```

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 10]
```diff
@@ -443,7 +471,11 @@ export abstract class ClientApplication {
             correlationId: correlationId
         };
         this.preflightBrowserEnvironmentCheck(InteractionType.Silent);
-        const ssoSilentMeasurement = this.performanceClient.startMeasurement(PerformanceEvents.SsoSilent, correlationId);
+        this.ssoSilentMeasurement = this.performanceClient.startMeasurement(PerformanceEvents.SsoSilent, correlationId);
+        this.ssoSilentMeasurement?.addStaticFields({
+            visibilityChange:false
```
Comment (tnorling) [inline lib/msal-browser/src/app/ClientApplication.ts:466]: Don't need to add this at all if it doesn't happen. In any case I don't think our pattern should involve overwriting static fields after they've already been assigned

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 11]
```diff
@@ -415,6 +418,31 @@ export abstract class ClientApplication {
         });
     }
 
+    private trackPageVisibilityWithMeasurement():void {
```
Comment (tnorling) [inline lib/msal-browser/src/app/ClientApplication.ts:422]: I wonder if this is something we can do in the performance client itself so we don't need to have logic dependent on the truthiness of the ssoSilent/acquireTokenByCode variables?

**Review by tnorling — CHANGES_REQUESTED** [review]

File: lib/msal-browser/src/app/ClientApplication.ts (hunk 6 above)
Comment (konstantin-msft) [reply lib/msal-browser/src/app/ClientApplication.ts]: fixed

File: lib/msal-browser/src/app/ClientApplication.ts (hunk 7 above)
Comment (konstantin-msft) [reply lib/msal-browser/src/app/ClientApplication.ts]: fixed

File: lib/msal-browser/src/app/ClientApplication.ts (hunk 8 above)
Comment (konstantin-msft) [reply lib/msal-browser/src/app/ClientApplication.ts:161]: The purpose is to be able to use `trackPageVisibilityWithMeasurement` function as an event listener. I have added a comment.

File: lib/msal-browser/src/app/ClientApplication.ts (hunk 11 above)
Comment (konstantin-msft) [reply lib/msal-browser/src/app/ClientApplication.ts:422]: I would rather keep this method here as it belongs to the browser lib only.

File: lib/msal-browser/src/app/PublicClientApplication.ts (hunk 4 above)
Comment (konstantin-msft) [reply lib/msal-browser/src/app/PublicClientApplication.ts]: fixed

File: lib/msal-browser/src/app/ClientApplication.ts (hunk 3 above)
Comment (konstantin-msft) [reply lib/msal-browser/src/app/ClientApplication.ts]: fixed

File: lib/msal-browser/src/app/ClientApplication.ts (hunk 2 above)
Comment (konstantin-msft) [reply lib/msal-browser/src/app/ClientApplication.ts]: fixed

File: change/@azure-msal-browser-59aee50c-dfa5-47b3-aba3-935707c03d9f.json (hunk 1 above)
Comment (konstantin-msft) [reply change/@azure-msal-browser-59aee50c-dfa5-47b3-aba3-935707c03d9f.json]: fixed

File: lib/msal-browser/src/app/ClientApplication.ts (hunk 9 above)
Comment (konstantin-msft) [reply lib/msal-browser/src/app/ClientApplication.ts]: fixed

File: lib/msal-browser/src/app/ClientApplication.ts (hunk 10 above)
Comment (konstantin-msft) [reply lib/msal-browser/src/app/ClientApplication.ts:466]: Looks like it's fixed already.

**Pull request closed without merging**